FHEM/14_SD_UT.pm

```diff
@@ -338,1656 +340,1656 @@ use warnings;
+        } elsif ($attrValue eq 'Techmar') {
+          $deviceCode = substr($bitData,0,32);
+          $deviceCode = sprintf("%08X", oct( "0b$deviceCode" ) );
+          $devicename = $devicemodel.'_'.$deviceCode;
+        ############ Navaris  ############
+        } elsif ($attrValue eq 'Navaris') {
+          $deviceCode = sprintf("%06X", oct( "0b$bitData" ) );
+          $devicename = $devicemodel.'_'.$deviceCode;
+        ############ unknown ############
+        } else {
+          $devicename = 'unknown_please_select_model';
+          Log3 $name, 3, "SD_UT_Attr UNDEFINED sensor $attrValue (model=unknown)";
+        }
+
+        Log3 $name, 3, "SD_UT_Attr UNDEFINED sensor $attrValue detected, code $deviceCode (DoTrigger)" if ($devicemodel ne 'unknown');
+
+        $modules{SD_UT}{defptr}{deletecache} = $name; # delete old device
+        Log3 $name, 5, "SD_UT: Attr cmd=$cmd devicename=$name attrName=$attrName attrValue=$attrValue oldmodel=$oldmodel";
+
+        readingsSingleUpdate($hash, 'state', $state, 0);
+
+        DoTrigger ('global','UNDEFINED unknown_please_select_model SD_UT unknown') if ($devicename eq 'unknown_please_select_model');   # if user push attr return to unknown
+        DoTrigger ('global',"UNDEFINED $devicename SD_UT $devicemodel $deviceCode") if ($devicename ne 'unknown_please_select_model');  # create new device
+
+        $attr{$devicename}{model} = $attrValue; # set model
+
+      } else {
+        readingsSingleUpdate($hash, 'state', 'Please press button again!', 0);
+        return "Please press button again or receive more messages!\nOnly with another message can the model be defined.\nWe need bitMSG from message.";
+      }
+    }
+  }
+
+  if ($cmd eq 'del' && $attrName eq 'model') {      ### delete readings
+    for my $readingname (qw/Button deviceCode LastAction state unknownMSG/) {
+      readingsDelete($hash,$readingname);
+    }
+  }
+
+  ## return if fhem init
+  if ($init_done) {
+    Log3 $name, 3, "SD_UT_Attr set $attrName to $attrValue" if ($cmd eq 'set');
+    return "Note: Your unknown_please_select_model device are deleted with the next receive.\nPlease use your new defined model device and do not forget to push -Save config-" if ($defs{$name}->{DEF} eq 'unknown' && $oldmodel eq 'unknown' && $attrValue ne "$oldmodel");
+  }
+  return;
 }
 
 ###################################
 sub SD_UT_bin2tristate($) {
-	my $bitData = shift;
-	my %bintotristate=(
-	 	 "00" => "0",
-	 	 "10" => "F",
-	 	 "11" => "1"
-	);
-	my $tscode;
-	for (my $n=0; $n < length($bitData); $n = $n + 2) {
-		$tscode = $tscode . $bintotristate{substr($bitData,$n,2)};
-	}
-	return $tscode;
+  my $bitData = shift;
+  my %bintotristate=(
+     '00' => '0',
+     '10' => 'F',
+     '11' => '1'
+  );
+  my $tscode;
+  for (my $n=0; $n < length($bitData); $n = $n + 2) {
+    $tscode = $tscode . $bintotristate{substr($bitData,$n,2)};
+  }
+  return $tscode;
 }
 
 ###################################
 sub SD_UT_tristate2bin($) {
```

Comment:
Durch 
`sub SD_UT_tristate2bin`
ersetzen.

Wobei ich jetzt tippen würde, dass wir diesen code schon mal in SD_Protocols haben und er hier verschwinden kann. ? Oder nur teilweise?

Reply:
So einfach geht das nicht. Ich hatte es so als erstes aber dann kommen Meldungen wie 
„Too many arguments for main:: ... „

Ob wir die Funktion aus SD_Protocols nutzen kann müsste ich schauen.

Reply:
Laut Commit History hattest Du `sub SD_UT_tristate2bin();` und dieser Prototyp erlaubt dann **keine** Variablen, aber mein Vorschlag und der sollte auch so aus perlcritic kommen bedeutet, dass es keinen Prototypen mehr gibt und die Validierung in der sub erfolgt.

Reply:
>Wobei ich jetzt tippen würde, dass wir diesen code schon mal in SD_Protocols haben und er hier verschwinden kann. ? Oder nur teilweise?

So nicht, nur anders und mit anderem Ergebnis. Zusätzlich ist nur eine von beiden Verfügbar, aber mit anderem Ergebnis zum Schluss. Ich würde diese dort im Modul belassen.